### PR TITLE
Initial draft of stake delegation without fees and penalties.

### PR DESCRIPTION
**Pull request by nikita-fuchs**
This PR has been superseded by [PR #39 ]

### CODE REVIEW

**Comment by nikita-fuchs** [comment]
@happi would you please assign yourself, piwo and hans as reviewers? I don't have the rights. thanks.

File: delegated-staking.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  
+  
+### **Contextual constraints**
+ The above delegation logic assumes the following requirements to the `main staking logic` to be met: 
+ 
+ 1. On any adjustment of a staker's amount, who registered as a `delegatee`, `update_delegatees_stake()` is called to keep the delegation logic's book-keeping in sync with the amount of funds staked by stakers in the `main staking logic`. By checking if the Caller is a `delegatee`, all delegatee's calls to adjust their stake shall be forwarded *to the delegation logic* (as for example the direct withdrawal from the staker's total stake in the `main staking logic` would withdraw also the funds *delegated* to him). The delegation logic has means of adjusting the staked amount in the `main staking logic` in accordance to the changes in its book keeping, according stubs are included in the current code.
+ 3. It is assumed there is some function in the `main staking logic` that is called by the block producer every time he produces a block. (Current information: There is supposedly a function called `step()` to which this applies).
```
Comment (hanssv) [inline delegated-staking.md:167]: We will not do this on a block-by-block basis...

File: delegated-staking.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+ 4. It is assumed that rewards for blocks produced by stakers that registered as `delegatee`s remain in the contract of the `main staking logic`, from which the eligible parties can withdraw them utilizing the delegation logic's `withdraw_rewards()`.
```
Comment (hanssv) [inline delegated-staking.md:168]: This is _one_ way a delegatee/validator might organize his/her "pool" other variants are feasible too, right?

File: delegated-staking.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+ 3. It is assumed there is some function in the `main staking logic` that is called by the block producer every time he produces a block. (Current information: There is supposedly a function called `step()` to which this applies).
+ 4. It is assumed that rewards for blocks produced by stakers that registered as `delegatee`s remain in the contract of the `main staking logic`, from which the eligible parties can withdraw them utilizing the delegation logic's `withdraw_rewards()`. 
+ 5. Every time a block was produced, the `split_reward()` function is to be called (presumably in the function referenced in 3.) . This is safe, because `split_reward()` immediately returns, if the caller is not a registered `delegatee`. It is assumed, that in this call, the earned block reward is `Call.value`. If that is not the case, the delegation logic can easily be adjusted to reference this value from somewhere else which is accessible to the contract.
```
Comment (hanssv) [inline delegated-staking.md:169]: At the end of an _epoch_ - and with some  (1 epoch?) delay - rewards are distributed. The `StakingValidator` interface allows a callback to be registered; this function will be called with information of rewards during the given epoch.

File: delegated-staking.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    
+2. Although `split_reward_to_delegators()` only distributes funds nominally instead of transfering actual value, it is to be verified whether the call to it has actually `value` assigned to it or not, therefore adding / removing the `payable` modifier might be necessary.
+  
+  
+### **Contextual constraints**
+ The above delegation logic assumes the following requirements to the `main staking logic` to be met: 
+ 
+ 1. On any adjustment of a staker's amount, who registered as a `delegatee`, `update_delegatees_stake()` is called to keep the delegation logic's book-keeping in sync with the amount of funds staked by stakers in the `main staking logic`. By checking if the Caller is a `delegatee`, all delegatee's calls to adjust their stake shall be forwarded *to the delegation logic* (as for example the direct withdrawal from the staker's total stake in the `main staking logic` would withdraw also the funds *delegated* to him). The delegation logic has means of adjusting the staked amount in the `main staking logic` in accordance to the changes in its book keeping, according stubs are included in the current code.
```
Comment (hanssv) [inline delegated-staking.md:166]: This is not settled, but we are leaning towards the main staking contract not knowing if the logic behind a `StakingValidator` is a single entity or a delegatee backed by several delegates.

**Comment by happi** [comment]
Hmm. Not sure how it got closed.

**Pull request closed without merging**